DirectXTex/DirectXTexImage.cpp

```diff
@@ -567,6 +567,78 @@ HRESULT ScratchImage::InitializeFromImage(const Image& srcImage, bool allow1D, C
     return S_OK;
 }
 
+_Use_decl_annotations_
+HRESULT ScratchImage::InitializeSwizzleFromImage(const Image& srcImage, CP_FLAGS flags) noexcept
+{
+    HRESULT hr = Initialize2D(srcImage.format, srcImage.width, srcImage.height, 1, 1, flags);
+
+    if (FAILED(hr))
+        return hr;
+
+    uint8_t* sptr = srcImage.pixels;
+    if (!sptr)
+        return E_POINTER;
+
+    uint8_t* dptr = m_image[0].pixels;
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
+
+    uint32_t xBytesMask = 0b1010101010101010;
+    uint32_t yBytesMask = 0b0101010101010101;
+
+    for (size_t y = 0; y < srcImage.height; y++)
+    {
+        for (size_t x = 0; x < srcImage.width; x++)
+        {
+            uint32_t swizzleIndex = _pdep_u32(x, xBytesMask) + _pdep_u32(y, yBytesMask);
+
+            uint8_t* sourcePixelPointer = sptr + y * srcImage.rowPitch + x * bytesPerPixel;
+            uint8_t* destPixelPointer = dptr + swizzleIndex * bytesPerPixel;
+            memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+        }
+    }
+
+    return S_OK;
+}
+
+_Use_decl_annotations_
+HRESULT ScratchImage::InitializeImageFromSwizzle(const Image& srcImage, CP_FLAGS flags) noexcept
+{
+    HRESULT hr = Initialize2D(srcImage.format, srcImage.width, srcImage.height, 1, 1, flags);
+
+    if (FAILED(hr))
+        return hr;
+
+    uint8_t* sptr = srcImage.pixels;
+    if (!sptr)
+        return E_POINTER;
+
+    uint8_t* dptr = m_image[0].pixels;
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
+
+    uint32_t xBytesMask = 0b1010101010101010;
```

Comment:
I believe the 3D case is to process the entire 'slice' at a time since they are continuous in memory and you use:

```
size_t swizzleIndex = _pdev_u32(x, maskx) + _pdep_u32(y, masky) + _pdep_u32(z, maskz);
```

Reply:
Regarding bit pattern versus bytes per pixel:

I also was confused by this.  The variable bit patterns are to be used with address offset.  Notice how they ignore the lower bits for alignment?

I'm using a bit pattern for index.  And then converting to bpp specific memory offset later.

Reply:
Note I had to edit the 3D patterns. My source encoded them slightly shifted for a LUT thing which doesn't apply to our usage.

Reply:
There are different patterns based on the BPP. The index is the index to the pixel regardless of its size, not the byte.

Note that this means the 'pitch' is ignored completely in the swizzled image and you can end up with some extra memory at the end for cases where pitch is not the same as ``bpp*w``. This space is zero filled for most cases, although apparently for BC they replicate the last block data instead.

Reply:
"The index is the index to the pixel regardless of its size"
Yes, that's why I'm swizzleing the index.  Because there's only one pattern.

"the 'pitch' is ignored completely in the swizzled image"
Yes, that's why .rowPitch is only used in rowMajorOffset.  Not is Swizzle[Index|Offset].

I don't see the problem.